managers/views.py

```diff
@@ -92,14 +92,10 @@ def get_success_url(self) -> str:
         return reverse("custom-project-detail", kwargs={"pk": self.kwargs["pk"]})
 
 
-class ProjectDeleteView(DeleteView):
-    model = Project
-
-    def dispatch(self, request: HttpRequest, *args: Any, **kwargs: Any) -> HttpResponse:
-        if not request.user.is_admin:
-            return redirect(reverse("home"))
-
-        return super().dispatch(request, *args, **kwargs)
-
-    def get_success_url(self) -> str:
-        return reverse("custom-projects-list")
+def delete_project(request, pk):
```

Comment:
How this change is fixing the bug?

Reply:
Basically, *DeleteView* requires template to work. It serves as a confirmation page (we implemented this functionality as popup in previous view). With no template provided, the app crashes. With any template provided, a *GET* request is returned after closing popup in previous view and thus the template is rendered with no deletion being performed (because delete view is still waiting for *POST* to do that). That's definitely not how it should work at all.

At the time I was supposed to apply the fix ASAP, so it was a lot quicker for me to rewrite it to function rather than tangle with Django's mysterious inner mechanics in it's high-level View classes.

It is doable of course, I'd guess we have to overwrite *get* method so it calls for deletion and then redirects to desired address.

However, from our personal experience (since I developed entirety of *eMunchkin* based on generic views on my side and I strongly urged others to do the same) it was a lot more time consuming to figure out how the generic views work to know where exactly we must apply changes in functionality that suit our needs (which granted, was a learning experience at the time), than just writing it from scratch, or at the very least from lower-level abstraction.

Since the DeleteView is designed to suit much more sophisticated function than we require, which is to delete given object upon call and redirect (it literally works like a basic function) I'd strongly encourage to leave it as a function-based view, due to it's simplicity. Remodelling inner workings of complex Django View to suit such simplistic function-like purpose, is a major overkill in my opinion.

I am well aware of benefits regarding the principle of keeping all views in same technology, but I'd argue this is a valid exception. I wouldn't use complex mechanics for simplified actions. I think it's an unnecessary waste of resources, both in time and code.

This is also the exact reason we had to write our own views derived from ApiView rather than using generic views when we were still working on REST framework.

Reply:
> Basically, DeleteView requires template to work. 

No, it depends if we want to use both `GET` and `POST`. Here we only use `POST`.

> With any template provided, a GET request is returned after closing popup in previous view and thus the template is rendered with no deletion being performed (because delete view is still waiting for POST to do that).

I honestly don't understand this part. Do you mean that the problem is that popup is making GET call or that the view is returning redirect to it? Because both can be changed either in popup itself, or in view - it can return redirect, template, or ajax response.

> At the time I was supposed to apply the fix ASAP, so it was a lot quicker for me to rewrite it to function rather than tangle with Django's mysterious inner mechanics in it's high-level View classes.

But it has nothing to do with the bug itself. The bug is about missing jquery-ui script that doesn't allow popup to open. The view is working OK.

> it was a lot more time consuming to figure out how the generic views work to know where exactly we must apply changes in functionality that suit our needs

Well, that the part of this project to get you guys familiar with how it works. And you don't actually have to do this, because it is already implemented, so you can just optionally get familiar with how it works.

> Remodelling inner workings of complex Django View to suit such simplistic function-like purpose, is a major overkill in my opinion.
> I wouldn't use complex mechanics for simplified actions.

They are simple generic views that covers most of the possible use cases with overriding 1-2 methods. Getting familiar with how they work and getting used to use them allows you to cover this cases with few lines of code.

>  I think it's an unnecessary waste of resources, both in time and code.

Again, this is already done.

Reply:
>I honestly don't understand this part. Do you mean that the problem is that popup is making GET call or that the view is returning redirect to it? Because both can be changed either in popup itself, or in view - it can return redirect, template, or ajax response.

Currently we are getting there by link in *<a>* tag. If a request method can be specified I wasn't aware of it (though it does seem obvious). By default, a GET was sent and the app crashed.

> But it has nothing to do with the bug itself. The bug is about missing jquery-ui script that doesn't allow popup to open. The view is working OK.

After I made the button work again, the app crashed upon call for deleting. Since those two are tightly related, I figured I could quickly fix this newfound bug in same Pull Request.

But since the only change I applied was moving the scripts to different place in code, I'm certain the DeleteView still wouldn't work, because the way it's called from *project_form* didn't change.

> Well, that the part of this project to get you guys familiar with how it works. And you don't actually have to do this, because it is already implemented, so you can just optionally get familiar with how it works.

If we treat it as a learning experience, sure. I'll be happy to dive into these mechanics and figure them out. I already had a good taste of it during *eMunchkin* development.
But recently, there is a lot of pressure regarding development, since we're struggling to deliver core functionality by the end of the month. This is why I was eager to pick faster solution that I'm familiar with, no to mention the time constraint at that particular time.

Reply:
Speaking of learning, I have a question: If you were to choose between generic view and function-based view in this case I take it you'd go for generic view. May I ask why is that specifically? I'd think that if the view's functionality is simple enough for a single, easily readable function, it'd be a better choice. I'd consider using complex class in such case to be at the very least a waste of memory. After all, when an instance is created, all of those unused features from parent class (like entirety of *GET* request handling) must be loaded, if I'm thinking correctly, making this view more resource consuming than it would be as a function. I am aware those differences are insignificant with today's tech, but to me it's a good design principle. Especially if function-based implementation would be faster to develop. Thus I'm wondering what are the benefits of deriving such view from generic class?

Reply:
@maciejSamerdak @kbeker @Szymiks read this too please.

> This is why I was eager to pick faster solution that I'm familiar with, no to mention the time constraint at that particular time.

Sure, that's understandable. If we have situation like this and you are waiting for me you may always merge current state when its working **but** please add an issue in github so we can remember to come back to this later.

>  I am aware those differences are insignificant with today's tech, but to me it's a good design principle. Especially if function-based implementation was faster to develop. Thus I'm wondering, what are the benefits of deriving such view from generic class?

About benefits:
- Once you know how to use particular class it is very simple to create a new subclass of it and replace only the parts that you need. When you write function based view you have to provide full implementation each time. This is especially good for those simple, **repeatable** views that Django provides as generics. So I can't agree that its faster once you learn it, I think its opposite.
- When you are subclassing a generic view from popular framework like Django you can be more sure that it is implemented without bugs and well tested. It is always easier to make a bug when you create your own implementation (I don't mean "you", I mean anyone). 
- You can use multiple inheritance. You can create your own generic classes which inherits from Django and does your custom logic and reuse them. You can use mixin classes. Basically you can use all the power of object oriented programming (which I hope we will have chance to use once we move all stuff to Django views).
- I think they are more readable. In example you can see you have a subclass of `TemplateView` with custom template as an attribute and you already know everything about what this view does.

> After all, when an instance is created, all of those unused features from parent class

Well this was actually a little untypical case that we wanted to use only `POST` part of this view, regularly you would use it fully. And if you need something in the middle you can always add it by just overriding some already existing method, without a need to modify whole function structure.

Those are just first things that comes to my mind. Off course this subject is probably better described in Django-related literature too. You can search for "class based views vs function based views django" and see different opinions and arguments.

> I'd consider using complex class in such case to be at the very least a waste of memory.

Its good that you think about things like this but I would not worry too much about it unless we get to that point. I mean, as you have said, with today's tech its rather unnoticeable in web project with this target usage. I think worse bottlenecks could be number/complexity of SQL queries, size of http response or static files.